health_vara_report.py

```diff
@@ -21,69 +34,64 @@ def execute(cls, ids, data):
-        finding_details = defaultdict(int)
-        lang = IrLang.get()
-        last_finding = None
-        for patient in records:
-            for finding in [f for f in patient.findings if f.bi_rads]:
-                if not last_finding:
-                    last_finding = finding
-                fields_with_content = [field for field in finding._fields
-                    if getattr(finding, field, None)
-                    and field in detail_fields]
-                if fields_with_content:
-                    item = 0
-                    finding_details[finding.id] = ''
-                    for field in fields_with_content:
-                        field_def = finding._fields[field].definition(
-                            finding, lang.code)
-                        # TODO: translate values
-                        value = getattr(finding, field, None)
-                        if field_def['type'] == 'boolean':
-                            if value:
-                                value = 'Yes'
-                            else:
-                                value = 'No'
-                        if item > 0:
-                            finding_details[finding.id] += '\n'
-                        label = format_label(field_def['string'])
-                        finding_details[finding.id] += '%s%s' % (label,
-                            value.capitalize())
-                        item += 1
-
-        context['company'] = company
-        context['report_address'] = report_address
+        # Note: `records` should always be a singleton list.
+        assert (len(records) == 1)
+        patient = records[0]
+
+        # Sort imaging test results for each patient ID in reverse chronological
+        # order. We need the most recent imaging results recorded for a patient.
+        imaging_results = sorted(patient.imaging_test_results,
+                                key=lambda r: r.create_date,
+                                reverse=True)
+
         context['today'] = Date.today()
-        context['finding_details'] = finding_details
-        context['last_finding'] = last_finding
+        context['has_ultrasound_findings'] = False
+
+        findings = []
+        imaging_result = None
+        assessment = None
+        density = None
+        last_finding = None
+
+        # Processing findings associated with the latest Imaging Result.
+        if len(imaging_results) != 0:
+            imaging_result = imaging_results[0]
+            findings = findings + list(imaging_result.findings)
+            density = imaging_result.density
+            assessment = imaging_result.assessment
+
+        # Processing findings outside the imaging results workflow.
+        imaging_finding_ids = set(map(lambda f: f.id, findings))
+        other_findings = list(filter(lambda f: f.id not in imaging_finding_ids,
+                                     patient.findings))
+        findings = findings + other_findings
```

Comment:
At this stage, do we have any guarantee that we have at least one finding?
I can't track it down but maybe it's a check from before this method?

We seem to be relying on it a bit when finding `last_finding` density on line 76.

Reply:
> At this stage, do we have any guarantee that we have at least one finding?

It's possible use this report with no findings. In that case, most of the values provided by the context are going to be `None`. We use `last_finding` in this method only if there is at least one item in `other_findings`.

Reply:
Ahh okay, empty lists in python are falsy, I forgot, all good!